### ISSUE TITLE
Upgrade daemon without restarting containers

### ISSUE THREAD

**Issue by shykes** [docker/docker#2658]
Docker needs a way to upgrade itself graciously, with minimum interruption of service. Ideally all containers would continue to function with zero downtime and zero behavior change. This might not always be possible, for example if the upgrade introduces significant changes to the container's runtime environment itself. In that case, docker should give the sysadmin maximum flexibility - in a perfect world the upgrade could be rolled out separately for each container.

**Comment by kim0** [comment]
Now that 1.5 is out .. can someone please hack on this so the world can update peacefully :)

**Comment by voycey** [comment]
This is the one thing currently that is preventing me from moving everything over :+1:

**Comment by bobrik** [comment]
Any update on this? Restarting all containers on a huge machine because `docker pull` hangs on small monitoring image doesn't look a like a good solution. Especially when you do it several times.

**Comment by mudverma** [comment]
Hi All,

I have put out a proposal for docker's hot upgrade and SPoF issue.  Please take a look. 

https://github.com/docker/docker/pull/13884

Thanks
Mudit

**Comment by calavera** [comment]
FWIW, this issue will be solved when #20662 lands in master.

**Comment by bobrik** [comment]
Is it really closed? https://github.com/docker/docker/pull/20662#issuecomment-197970944

**Comment by dvorak** [comment]
#20662 Doesn't appear to actually address this issue.  Can this be reopened?

**Comment by ayanamist** [comment]
This issue still exists. When i run some container like `docker run -d alpine tail -f /dev/null` and then restart docker daemon, the container disappears.
Tested on docker 1.11
Plz reopen the issue.

**Comment by thaJeztah** [comment]
Ah, yes, I think this should not be closed (yet); the groundwork for this feature has been laid in https://github.com/docker/docker/issues/20662, but it's not enabled yet (it will be available as an experimental feature first). I'll reopen until it's enabled by default

**Comment by dvorak** [comment]
@thaJeztah Can you explain at a high-level how this will work when it's in place?

**Comment by thaJeztah** [comment]
@dvorak basically, the `containerd-shim` allows both the `docker` daemon, and the `containerd` daemon to be restarted, without killing the container. Upon restart, the daemons can re-connect to the container, by use of the `containerd-shim`'s.

You can already test the basics, by installing an experimental version (https://experimental.docker.com), and running the daemon manually (we're investigating an issue if it's started through systemd https://github.com/docker/docker/issues/21933).

Basically;

    #manually start the daemon
    docker daemon -D &

Then, start a container, and `kill -9` the daemon. The container should remain running, but `docker ps` should give an error, because the daemon isn't running.

After that, starting the daemon again (`docker daemon -D &`), should re-attach it to the containers and `docker ps` should show the container.

**Comment by erich-comsIO** [comment]
Just issued a docker-machine upgrade on my Mac - destroyed all of the containers, which only represented about twelve hours of work. Is that supposed to happen?

**Comment by mbentley** [comment]
That's not really relevant to this issue. I would suggest reporting that here: https://github.com/docker/machine/issues

**Comment by erich-comsIO** [comment]
Sorry! Thought this was the closest hit… Appreciate reply.

**Comment by crosbymichael** [comment]
Added this to the 1.12 milestone and assigned to myself.  

What we currently have left to do is:

- [ ] Persist graph driver reference counting
- [ ] Persist network information for restore

**Comment by BSWANG** [comment]
/var/run/docker.sock mount in containers refuse connection after daemon hot upgrade https://github.com/docker/docker/issues/22789

**Comment by KramKroc** [comment]
Hi, I took the experimental release today to test this out today. I found a few things that I thought would be worth flagging (just to note that my containers were initially started with docker-compose):

1) After the daemon was restarted, the docker-proxy processes were killed and so could no longer server traffic to the exposed ports. the container itself seems to be fine, but without the port it's hard to send traffic to it to verify its operating as expected.

2) containers that I had previously stopped prior to killing and restarting the docker daemon were restarted. Will this be expected behaviour in final solution or will the container state be honoured after daemon restart?
```
$ docker info
Containers: 15
 Running: 13
 Paused: 0
 Stopped: 2
Images: 16
Server Version: 1.12.0-dev
Storage Driver: devicemapper
 Pool Name: docker-253:0-1567975-pool
 Pool Blocksize: 65.54 kB
 Base Device Size: 10.74 GB
 Backing Filesystem: xfs
 Data file: /dev/loop0
 Metadata file: /dev/loop1
 Data Space Used: 3.6 GB
 Data Space Total: 107.4 GB
 Data Space Available: 28.54 GB
 Metadata Space Used: 7.041 MB
 Metadata Space Total: 2.147 GB
 Metadata Space Available: 2.14 GB
 Thin Pool Minimum Free Space: 10.74 GB
 Udev Sync Supported: true
 Deferred Removal Enabled: false
 Deferred Deletion Enabled: false
 Deferred Deleted Device Count: 0
 Data loop file: /var/lib/docker/devicemapper/devicemapper/data
 WARNING: Usage of loopback devices is strongly discouraged for production use. Either use `--storage-opt dm.thinpooldev` or use `--storage-opt dm.no_warn_on_loop_devices=true` to suppress this warning.
 Metadata loop file: /var/lib/docker/devicemapper/devicemapper/metadata
 Library Version: 1.02.107-RHEL7 (2015-10-14)
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins: 
 Volume: local
 Network: null host bridge
Kernel Version: 3.10.0-327.el7.x86_64
Operating System: CentOS Linux 7 (Core)
OSType: linux
Architecture: x86_64
CPUs: 1
Total Memory: 7.64 GiB
Name: localhost.localdomain
ID: E33U:MJER:UUZR:LFBN:RIFS:KAP4:FP5F:OAQF:73AR:3GWU:LX5I:BC76
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 85
 Goroutines: 123
 System Time: 2016-05-24T12:13:22.032865741+01:00
 EventsListeners: 0
Username: 
Registry: https://index.docker.io/v1/
WARNING: bridge-nf-call-iptables is disabled
WARNING: bridge-nf-call-ip6tables is disabled
Experimental: true
Insecure Registries:
 127.0.0.0/8
```

**Comment by aluminous** [comment]
Hi, sorry to drag up an old issue, but did this make it into the 1.12.0 release or was it pushed back?

We recently upgraded, but restarting dockerd still stops running containers. If I understood correctly, killing dockerd shouldn't kill running containers anymore right?

**Comment by thaJeztah** [comment]
@aluminous yes, it was implemented in https://github.com/docker/docker/pull/23213, which is in 1.12.0, but it's not enabled by default. Documentation was added through https://github.com/docker/docker/pull/24970, and can be found here; https://docs.docker.com/engine/admin/live-restore/

**Comment by aluminous** [comment]
Apologies, I missed that section of the documentation. Thank you!

**Comment by thaJeztah** [comment]
No worries, I noticed it wasn't referred to from this issue, so was good to add some links here for reference anyway :smile:
